quic/s2n-quic-platform/src/message/cmsg.rs

```diff
@@ -229,87 +231,111 @@ pub fn decode(msghdr: &libc::msghdr) -> AncillaryData {
 /// # Safety
 ///
 /// `cmsghdr` must refer to a cmsg containing a payload of type `T`
-unsafe fn decode_value<T: Copy>(cmsghdr: &libc::cmsghdr) -> T {
-    use core::{mem, ptr};
+#[inline]
+unsafe fn decode_value<T: Copy>(value: &[u8]) -> T {
+    use core::mem;
 
-    assert!(mem::align_of::<T>() <= mem::align_of::<libc::cmsghdr>());
-    debug_assert_eq!(
-        cmsghdr.cmsg_len as usize,
-        libc::CMSG_LEN(mem::size_of::<T>() as _) as usize
-    );
-    ptr::read(libc::CMSG_DATA(cmsghdr) as *const T)
+    debug_assert!(mem::align_of::<T>() <= mem::align_of::<cmsghdr>());
+    debug_assert!(value.len() >= size_of::<T>());
+
+    let mut v = mem::zeroed::<T>();
+
+    core::ptr::copy_nonoverlapping(value.as_ptr(), &mut v as *mut T as *mut u8, size_of::<T>());
+
+    v
 }
 
 struct Iter<'a> {
-    msghdr: &'a libc::msghdr,
-    cmsghdr: Option<&'a libc::cmsghdr>,
+    cursor: *const u8,
+    end_ptr: *const u8,
+    contents: core::marker::PhantomData<&'a [u8]>,
 }
 
 impl<'a> Iter<'a> {
     /// Creates a new cmsg::Iter used for iterating over control message headers in the given
-    /// `msghdr` using the `CMSG_FIRSTHDR` and `CMSG_NXTHDR` macros.
+    /// `msghdr`.
     ///
     /// # Safety
     ///
-    /// `msghdr` must contain valid control messages and be readable for the lifetime
-    /// of the returned `Iter`
-    unsafe fn new(msghdr: &'a libc::msghdr) -> Self {
+    /// * `contents` must contain valid control messages and be readable for the lifetime
```

Comment:
I think readability is guaranteed as the returned iter has the same lifetime? I would probably write the function as `-> Iter<'_>` or even explicitly with 'a to make sure elision does the right thing here and it's clearer for readers.

Reply:
You're right. Some of the phrasing is left over from the previous one and doesn't really make sense now.

Reply:
updated